File: src/typescript-port/id_pm.ts
// ID_PM.TS
// Ported from ID_PM.C - Page Manager
// Loads VSWAP file and provides page-based access to wall textures, sprites, and sound pages.

export const PMPageSize = 4096;

export interface PageListStruct {
    offset: number;
    length: number;
    locked: number;     // PMLockType: 0=unlocked, 1=locked
    mainPage: number;
    lastHit: number;
}

export let ChunksInFile = 0;
export let PMSpriteStart = 0;
export let PMSoundStart = 0;
export let PMPages: PageListStruct[] = [];

let pageData: (Uint8Array | null)[] = [];
let vswapData: ArrayBuffer | null = null;

export let PageFileName = 'VSWAP.';

//===========================================================================
// PM_Startup - Loads the VSWAP file
//===========================================================================

export async function PM_Startup(): Promise<void> {
    const extension = 'WL6';
    const filename = PageFileName + extension;

    const response = await fetch(`/${filename}`);
    if (!response.ok) {
        throw new Error(`PM_Startup: Could not load ${filename}`);
    }

    vswapData = await response.arrayBuffer();
    const view = new DataView(vswapData);

    // VSWAP header:
    // Uint16: ChunksInFile
    // Uint16: PMSpriteStart
    // Uint16: PMSoundStart
    ChunksInFile = view.getUint16(0, true);
    PMSpriteStart = view.getUint16(2, true);
    PMSoundStart = view.getUint16(4, true);

    // Then ChunksInFile Uint32 page offsets, then ChunksInFile Uint16 page lengths
    const headerSize = 6;
    PMPages = [];
    pageData = [];

    for (let i = 0; i < ChunksInFile; i++) {
        const offset = view.getUint32(headerSize + i * 4, true);
        const length = view.getUint16(headerSize + ChunksInFile * 4 + i * 2, true);
        PMPages.push({
            offset,
            length,
            locked: 0,
            mainPage: -1,
            lastHit: 0,
        });
        pageData.push(null);
    }
}

//===========================================================================
// PM_Shutdown
//===========================================================================

export function PM_Shutdown(): void {
    pageData = [];
    PMPages = [];
    vswapData = null;
}

//===========================================================================
// PM_GetPage - Returns a page's data
//===========================================================================

export function PM_GetPage(pagenum: number): Uint8Array | null {
    if (pagenum < 0 || pagenum >= ChunksInFile) return null;

    if (!pageData[pagenum]) {
        // Load page on demand
        const page = PMPages[pagenum];
        if (!vswapData || page.offset === 0 || page.length === 0) return null;

        pageData[pagenum] = new Uint8Array(vswapData, page.offset, page.length);
    }

    PMPages[pagenum].lastHit++;
    return pageData[pagenum];
}

//===========================================================================
// PM_GetPageAddress - same as PM_GetPage for our purposes
//===========================================================================

export function PM_GetPageAddress(pagenum: number): Uint8Array | null {
    return pageData[pagenum] || null;
}

//===========================================================================
// Convenience functions matching C macros
//===========================================================================

export function PM_GetSoundPage(v: number): Uint8Array | null {
    return PM_GetPage(PMSoundStart + v);
}

export function PM_GetSpritePage(v: number): Uint8Array | null {
    return PM_GetPage(PMSpriteStart + v);
}

//===========================================================================
// Other PM functions
//===========================================================================

export function PM_Reset(): void {
    // Reload all pages
    for (let i = 0; i < pageData.length; i++) {
        if (PMPages[i].locked === 0) {
            pageData[i] = null;
        }
    }
}

export function PM_Preload(_update: (current: number, total: number) => boolean): void {
    // Preload all pages
    for (let i = 0; i < ChunksInFile; i++) {
        PM_GetPage(i);
    }
}

export function PM_NextFrame(): void {
    // No-op in browser
}

export function PM_SetPageLock(pagenum: number, lock: number): void {
    if (pagenum >= 0 && pagenum < ChunksInFile)
        PMPages[pagenum].locked = lock;
}

export function PM_SetMainMemPurge(_level: number): void {
    // No-op
}

export function PM_CheckMainMem(): void {
    // No-op
}

export function PM_SetMainPurge(_level: number): void {
    // No-op
}

File: src/csharp-sdl3-port/IdPm.cs
// ID_PM.C -> IdPm.cs
// Page Manager - loads all VSWAP pages into memory at startup

using System;
using System.IO;

namespace Wolf3D
{
    public static class IdPm
    {
        private static bool PMStarted;
        private static int PMFrameCount;
        private static bool PMPanicMode;
        private static bool PMThrashing;

        public static void PM_Startup()
        {
            if (PMStarted) return;

            PML_OpenPageFile();
            PMStarted = true;
        }

        public static void PM_Shutdown()
        {
            if (!PMStarted) return;
            PMStarted = false;
            WL_Globals.PMPages = null;
            WL_Globals.PMPageData = null;
        }

        private static void PML_OpenPageFile()
        {
            string filename = WL_Globals.PageFileName + WL_Globals.extension;
            if (!File.Exists(filename))
            {
                // Try without extension
                filename = WL_Globals.PageFileName;
                if (!File.Exists(filename))
                {
                    // Try various paths
                    string[] paths = { ".", "assets", "../assets", "../../assets" };
                    bool found = false;
                    foreach (string path in paths)
                    {
                        string test = Path.Combine(path, WL_Globals.PageFileName + WL_Globals.extension);
                        if (File.Exists(test)) { filename = test; found = true; break; }
                        test = Path.Combine(path, "VSWAP.WL6");
                        if (File.Exists(test)) { filename = test; found = true; break; }
                    }
                    if (!found)
                        WlMain.Quit("PML_OpenPageFile: Unable to open page file");
                }
            }

            using (var fs = File.OpenRead(filename))
            using (var br = new BinaryReader(fs))
            {
                // Read header
                WL_Globals.ChunksInFile = br.ReadUInt16();
                WL_Globals.PMSpriteStart = br.ReadUInt16();
                WL_Globals.PMSoundStart = br.ReadUInt16();

                int numChunks = WL_Globals.ChunksInFile;

                // Read offsets
                uint[] offsets = new uint[numChunks];
                for (int i = 0; i < numChunks; i++)
                    offsets[i] = br.ReadUInt32();

                // Read lengths
                ushort[] lengths = new ushort[numChunks];
                for (int i = 0; i < numChunks; i++)
                    lengths[i] = br.ReadUInt16();

                // Allocate page structures
                WL_Globals.PMPages = new PageListStruct[numChunks];
                WL_Globals.PMPageData = new byte[numChunks][];

                for (int i = 0; i < numChunks; i++)
                {
                    WL_Globals.PMPages[i] = new PageListStruct
                    {
                        offset = offsets[i],
                        length = lengths[i],
                        locked = PMLockType.pml_Unlocked,
                        mainPage = -1,
                        emsPage = -1,
                        xmsPage = -1,
                        lastHit = 0
                    };

                    if (offsets[i] != 0 && lengths[i] != 0)
                    {
                        fs.Seek(offsets[i], SeekOrigin.Begin);
                        WL_Globals.PMPageData[i] = br.ReadBytes(lengths[i]);
                        WL_Globals.PMPages[i].mainPage = 0;
                    }
                }
            }
        }

        public static byte[] PM_GetPageAddress(int pagenum)
        {
            if (pagenum < 0 || pagenum >= WL_Globals.ChunksInFile)
                return null;
            return WL_Globals.PMPageData[pagenum];
        }

        public static byte[] PM_GetPage(int pagenum)
        {
            if (pagenum < 0 || pagenum >= WL_Globals.ChunksInFile)
                WlMain.Quit("PM_GetPage: Invalid page request");

            byte[] result = WL_Globals.PMPageData[pagenum];
            if (result == null)
                WlMain.Quit("PM_GetPage: Page not in memory!");

            WL_Globals.PMPages[pagenum].lastHit = (uint)PMFrameCount;
            return result;
        }

        public static byte[] PM_GetSoundPage(int v)
        {
            return PM_GetPage(WL_Globals.PMSoundStart + v);
        }

        public static byte[] PM_GetSpritePage(int v)
        {
            return PM_GetPage(WL_Globals.PMSpriteStart + v);
        }

        public static void PM_SetPageLock(int pagenum, PMLockType lockType)
        {
            if (pagenum >= 0 && pagenum < WL_Globals.ChunksInFile)
                WL_Globals.PMPages[pagenum].locked = lockType;
        }

        public static void PM_NextFrame()
        {
            PMFrameCount++;
            if (PMFrameCount >= int.MaxValue - 4)
            {
                for (int i = 0; i < WL_Globals.ChunksInFile; i++)
                    WL_Globals.PMPages[i].lastHit = 0;
                PMFrameCount = 0;
            }
        }

        public static void PM_Reset()
        {
            PMPanicMode = false;
        }

        public static void PM_Preload(Func<int, int, bool> update)
        {
            update?.Invoke(1, 1);
        }

        public static void PM_SetMainMemPurge(int level)
        {
            // No-op - all pages always in memory
        }

        public static void PM_SetMainPurge(int level)
        {
            // No-op
        }

        public static void PM_CheckMainMem()
        {
            // No-op
        }
    }
}
